Issue 714 patch hybrid

misi opened the pull request:
I added two things.

Commit:
OAuth "alg" related consideration

Commit:
Fix macKey wasn't encoded in base64

taylor-b committed:
Minor editorial changes

taylor-b commented on webrtc.html:
```diff
@@ -241,8 +241,52 @@
+                <p>If the <a>Auth Server</a> doesn't have prior knowledge of the
+                capabilities of the client, then the <code>OAuth Client</code>
+                needs to provide information about the <a>ICE Agent</a> HMAC
+                "alg" capabilities. This information helps the <a>Auth
```
Doesn't this conflict with "the `alg` should always be `HS256`", as mentioned below?

misi replied:
(First of all sorry for this "raw" PR. My son broke Monday his arm badly and I am pressed for time.)
You are right. It is confusing. 
I should state in second half that this is the how it is stated in the PoP key exchange.
I want to tell here that if the Auth Server has prior knowledge about ICE Agent capability, then we do not need to send "alg" during OAuth communication.

taylor-b replied:
Ok, I think I understand. I'll try to make it more clear what text is talking about the general PoP key exchange process, and what text is WebRTC-specific.

taylor-b commented on webrtc.html:
```diff
@@ -241,8 +241,52 @@
+                </p>
+                <div class="note">
+                  <p>When [[!RFC7635]] is used in WebRTC context, this
+                  specification adds one additional consideration to [[!RFC7635]].
```
Aren't two considerations listed below?

misi replied:
I mean the key length must be 256 this was the main point,
the second that we use alg: HS256 is only the result of this. 

But I don't have strong position, if it is confusing then we should change..

taylor-b commented on webrtc.html:
```diff
@@ -241,8 +241,52 @@
+                  specification adds one additional consideration to [[!RFC7635]].
+                  </p>
+                  <p>This specification considers to use a constant,
+                  <code>256</code> bit HMAC key length.
```
Isn't this already stated above?

taylor-b commented on webrtc.html:
```diff
@@ -241,8 +241,52 @@
+                </div>
+                <p>
+                If a shorter HMAC key is needed by the <a>ICE Agent</a>, then it
+                MUST be truncated by the implementation.
```
Is it simply truncated? RFC7635 says the client "calculates a 160-bit key for HMAC-SHA-1 using SHA1 and taking the 256-bit key as input." Or am I confusing two different things? Is this documented anywhere besides the appendix?

misi replied:
This is a divergence from the IETF defined path too. 
During WG meeting when 256 bit was decided AFAIU was consensus to truncate back only, but you are right RFC7635 appendix example defines differently.

taylor-b replied:
This seems like something that needs to be standardized on the IETF side. The WebRTC TURN client and non-WebRTC TURN server need to agree on the key used for the MESSAGE-INTEGRITY calculation, correct?. I find it odd that this is only mentioned in the appendix of RFC7635, but it's the closest we have. @juberti, thoughts? Should we revisit the decision to simply truncate the key?

misi replied:
> The WebRTC TURN client and non-WebRTC TURN server need to agree on the key

No! No negotiation happens! Client sends the HMAC key to Server. (So Server receive this key from the client.)
Client sends the HMAC key in the encrypted body of the access_token.  Sever extracts from the access_token the HMAC key and verifies the Message Integrity.

misi replied:
>@juberti, thoughts? 

I think he presented during that decision. But I am not totally sure.

>Should we revisit the decision to simply truncate the key?

Yes I think this will be the smartest what we can do. 
Let's raise it again to make it sure, that everybody has the same understanding.

May worth to double check the "alg" so ice agent capability discovery topic too, 
because I am not sure I have presented the problem correctly for the first time.

taylor-b replied:
> No! No negotiation happens! Client sends the HMAC key to Server. (So Server receive this key from the client.)

I worded that poorly. They don't negotiate the key, but what I meant is that they need to use the same key to calculate the HMAC of the allocate request, correct?

> Client sends the HMAC key in the encrypted body of the access_token. Sever extracts from the access_token the HMAC key and verifies the Message Integrity.

Wouldn't the HMAC key in the access token still be 256 bits, since it's in the encrypted part of the access token? So, the server needs to shorten the 256-bit key to 160 bits to verify the MESSAGE-INTEGRITY. If it shortens the key by "using SHA1 and taking the 256-bit key as input", but the client truncated it, then the message integrity check will fail.

I'll add this to the virtual interim topics, assuming my understanding is correct.

misi replied:
> I meant is that they need to use the same key to calculate the HMAC of the allocate request, correct?

-  They need to have the same key
or
- They need to calculate  the shorter key using the same algorithm.

I missed this problem but your point is absolutely valid, that we can't truncate on sender side, if the other end use other algorithm to make the key shorter.
If we want to make it fast then we have to accept RFC7635 algorithm, and just apply it.

>Wouldn't the HMAC key in the access token still be 256 bits, since it's in the encrypted part of the access token? So, the server needs to shorten the 256-bit key to 160 bits to verify the MESSAGE-INTEGRITY. If it shortens the key by "using SHA1 and taking the 256-bit key as input", but the client truncated it, then the message integrity check will fail.

- RFC7635 does not define clearly the supported "alg"-s. 
Although that POP-KEY-DISTRIBUTION Section 6. propose that:
"This document does not register a new value for the OAuth Access
   Token Types registry nor does it define values to be used for the
   'alg' parameter since this is the responsibility of specifications
   defining the mechanism for clients interacting with resource servers."
- RFC7635 also do not define clearly in main body howto calculate shorter keys, only as you mentioned it is in Appendix B. 

> I'll add this to the virtual interim topics, assuming my understanding is correct.

We are on the same page. I agree with your points.

misi replied:
I think we should declare in our spec that we follow and use the algorithm that presented in RFC7635 Appendix B. to calculate key with SHA1 for SHA1.

According Actual IETF RFC, STUN/TURN supports SHA1 HMAC only. 
In STUN-bis SHA256 (with optional hash truncation) and SHA1 is presented.
In STUN-BIS Section 15.3. describes the rolling upgrade plan, or agility plan.

RFC 7635 using in the example HMAC-SHA-256-128. 
That also means that a 256 bit key needed as input (According rfc4868 Section 2.6)
So HMAC-SHA-256-128 and HS256 is telling the same for Authorization Server the client needs a 256 bit key for HMAC calculation.
Why HS256? 
JWK format must be used according PoP. 
In JWK it is more conform the HS256 usage because it is member of the in the IANA "JSON Web Signature and Encryption Algorithms" registry.

Should the W3C WebRTC spec manage somehow STUN-bis rolling update/agility plan? 
My guess that such transition is a long term change, and so Next WebRTC spec should address this issue.

taylor-b commented on webrtc.html:
```diff
@@ -241,8 +241,52 @@
+                  <p>This specification considers to use a constant,
+                  <code>256</code> bit HMAC key length.
+                  </p>
+                  <p>This specification proposes to set the <a>OAuth Client</a>
```
Could this be worded as "The OAuthClient SHOULD obtain the mac_key by requesting an alg value of HS256. This will result in a 256-bit HMAC key."?

misi replied:
Sure. Thanks!

taylor-b commented on webrtc.html:
```diff
@@ -241,8 +241,52 @@
+                approprite key length and algorithm type (e.g.
+                symmetric/asymmetric).
+                </p>
+                <p>According to [[!RFC7635]] Section 4.1, the HMAC key MUST be a
```
This could be a "note"

misi replied:
Ok

taylor-b commented on webrtc.html:
```diff
@@ -241,8 +241,52 @@
+                <p>The <dfn>OAuth Client</dfn> sends an <code>OAuth
+                Request</code> to the <a>Auth Server</a> with OAuth param
+                <code>alg</code> (The alg value is set according the
+                supported HMAC capabilities of the <a>ICE Agent</a>.) and
```
Again; isn't "alg" always set to HS256?

misi replied:
Yes, you are right.
Probably I wrote it because I am not happy and satisfied with WG decision to not discover alg from ICE Agent(STUN/TURN client). 

I try to tell both the WG decision, and also what normally should do according the IETF.
To give information about how the WG not go along the IETF path.

Alg value was decided to set HS256 because:
- Actually TURN HMAC supports only SHA1 and  SHA256. This are SHA family algs only.
- HMAC key MUST be symmetric as defined in RFC7635
- 256 bit key length was the decision of the WebRTC WG

But what if later new message integrity algorithm will be supported in TURN later on? 
e.g. Elliptic Curve ES256.
I am not sure that WebRTC WG considered this..
https://tools.ietf.org/html/rfc7518#section-3.1
So not HS256 but ES256.

All in all, I want to document what is the normal path and how the WebRTC WG simplified it.

taylor-b commented on webrtc.html:
```diff
@@ -241,8 +241,52 @@
+                <code>key</code>, <code>kid</code>, and further OAuth related
+                parameters.
+                </p>
+                <p>The length of the HMAC key SHOULD be considered as a static
```
I don't like the wording "considered as", or "considers to use" below. Is there different phrasing that could be used?

Commit:
Updated Version

 - add definitions for OAuth Client and Authorization Server
 - resons behind HS256 has been picked.
 - replace key truncation with SHA1 according RFC7635 Appendix B.
 - higlight more some parts.

taylor-b commented:
I'm going to go ahead and merge this into my PR, and make some editorial changes.
